esmvalcore/preprocessor/_area.py

```diff
@@ -440,108 +456,156 @@ def _correct_coords_from_shapefile(cube, cmor_coords, pad_north_pole,
     return lon, lat
 
 
-def _get_masks_from_geometries(geometries,
-                               lon,
-                               lat,
-                               method='contains',
-                               decomposed=False,
-                               ids=None):
-
-    if method not in {'contains', 'representative'}:
-        raise ValueError(
-            "Invalid value for `method`. Choose from 'contains', ",
-            "'representative'.")
+def _process_ids(
+    geometries: fiona.collection.Collection,
+    ids: list | dict | None,
+) -> tuple:
+    """Read requested IDs and ID keys."""
+    # If ids is a dict, it needs to have length 1 and all geometries needs to
+    # have the requested attribute key
+    if isinstance(ids, dict):
+        if len(ids) != 1:
+            raise ValueError(
+                f"If `ids` is given as dict, it needs exactly one entry, got "
+                f"{ids}"
+            )
+        key = list(ids.keys())[0]
+        for geometry in geometries:
+            if key not in geometry['properties']:
+                raise ValueError(
+                    f"Geometry {dict(geometry['properties'])} does not have "
+                    f"requested attribute {key}"
+                )
+        id_keys: tuple[str, ...] = (key, )
+        ids = ids[key]
+
+    # Otherwise, use SHAPE_ID_KEYS to get ID
+    else:
+        id_keys = SHAPE_ID_KEYS
 
-    selections = dict()
-    if ids:
+    # IDs should be strings or None
+    if not ids:
+        ids = None
+    if ids is not None:
         ids = [str(id_) for id_ in ids]
-    for i, item in enumerate(geometries):
-        for id_prop in SHAPE_ID_KEYS:
-            if id_prop in item['properties']:
-                id_ = str(item['properties'][id_prop])
+
+    return (id_keys, ids)
+
+
+def _get_requested_geometries(
+    geometries: fiona.collection.Collection,
```

Comment:
do we really want to hardassign types that may change in the future? ie fiona may change the name of `collection` without changing the actual API

Reply:
Good point, will remove 👍